docs/config.md
| 配置项                    | 描述                                                                                                                                                    | 默认值               |
|:-----------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------|
| open_driver            | 开启浏览器运行，若更新无数据可开启此模式，较消耗性能                                                                                                                            | False             |
| open_empty_category    | 开启无结果频道分类，自动归类至底部                                                                                                                                     | False             |
| open_filter_resolution | 开启分辨率过滤，低于最小分辨率（min_resolution）的接口将会被过滤                                                                                                               | True              |
| open_filter_speed      | 开启速率过滤，低于最小速率（min_speed）的接口将会被过滤                                                                                                                      | True              |
| open_hotel             | 开启酒店源功能，关闭后所有酒店源工作模式都将关闭                                                                                                                              | True              |
| open_hotel_foodie      | 开启 Foodie 酒店源工作模式                                                                                                                                     | True              |
| open_hotel_fofa        | 开启 FOFA、ZoomEye 酒店源工作模式                                                                                                                               | True              |
| open_keep_all          | 开启保留所有检索结果，会保留非模板频道名称的结果，推荐手动维护时开启                                                                                                                    | False             |
| open_m3u_result        | 开启转换生成 m3u 文件类型结果链接，支持显示频道图标                                                                                                                          | True              |
| open_multicast         | 开启组播源功能，关闭后所有组播源工作模式都将关闭                                                                                                                              | True              |
| open_multicast_foodie  | 开启 Foodie 组播源工作模式                                                                                                                                     | True              |
| open_multicast_fofa    | 开启 FOFA 组播源工作模式                                                                                                                                       | True              |
| open_online_search     | 开启关键字搜索源功能                                                                                                                                            | False             |
| open_proxy             | 开启代理，自动获取免费可用代理，若更新无数据可开启此模式                                                                                                                          | False             |
| open_request           | 开启查询请求，数据来源于网络（仅针对酒店源与组播源）                                                                                                                            | False             |
| open_service           | 开启页面服务，用于控制是否启动结果页面服务；如果使用青龙等平台部署，有专门设定的定时任务，需要更新完成后停止运行，可以关闭该功能                                                                                      | True              |
| open_sort              | 开启排序功能（响应速度、日期、分辨率）                                                                                                                                   | True              |
| open_subscribe         | 开启订阅源功能                                                                                                                                               | False             |
| open_update            | 开启更新，用于控制是否更新接口，若关闭则所有工作模式（获取接口和测速）均停止                                                                                                                | True              |
| open_update_time       | 开启显示更新时间                                                                                                                                              | True              |
| open_url_info          | 开启显示接口说明信息，用于控制是否显示分辨率、接口协议类型等信息，为$符号后的内容，播放软件使用该信息对接口进行描述                                                                                            | True              |
| open_use_cache         | 开启使用本地缓存数据，适用于查询请求失败场景（仅针对酒店源与组播源）                                                                                                                    | True              |
| open_use_old_result    | 开启使用历史更新结果（包含模板与结果文件的接口），合并至本次更新中                                                                                                                     | True              |
| app_port               | 页面服务端口，用于控制页面服务的端口号                                                                                                                                   | 8000              |
| final_file             | 生成结果文件路径                                                                                                                                              | output/result.txt |
| hotel_num              | 结果中偏好的酒店源接口数量                                                                                                                                         | 10                |
| hotel_page_num         | 酒店地区获取分页数量                                                                                                                                            | 1                 |
| hotel_region_list      | 酒店源地区列表，"全部"表示所有地区                                                                                                                                    | 全部                |
| ipv4_num               | 结果中偏好的 IPv4 接口数量                                                                                                                                      | 5                 |
| ipv6_num               | 结果中偏好的 IPv6 接口数量                                                                                                                                      | 5                 |
| ipv6_support           | 强制认为当前网络支持IPv6，跳过检测                                                                                                                                   | False             |
| ipv_type               | 生成结果中接口的协议类型，可选值：ipv4、ipv6、全部、all                                                                                                                     | 全部                |
| ipv_type_prefer        | 接口协议类型偏好，优先将该类型的接口排在结果前面，可选值：IPv4、IPv6、自动、auto                                                                                                        | 自动                |
| min_resolution         | 接口最小分辨率，需要开启 open_filter_resolution 才能生效                                                                                                              | 1920x1080         |
| min_speed              | 接口最小速率（单位M/s），需要开启 open_filter_speed 才能生效                                                                                                             | 0.2               |
| multicast_num          | 结果中偏好的组播源接口数量                                                                                                                                         | 10                |
| multicast_page_num     | 组播地区获取分页数量                                                                                                                                            | 1                 |
| multicast_region_list  | 组播源地区列表，"全部"表示所有地区                                                                                                                                    | 全部                |
| online_search_num      | 结果中偏好的关键字搜索接口数量                                                                                                                                       | 0                 |
| online_search_page_num | 关键字搜索频道获取分页数量                                                                                                                                         | 1                 |
| origin_type_prefer     | 结果偏好的接口来源，结果优先按该顺序进行排序，逗号分隔，例如：hotel,multicast,subscribe,online_search；hotel：酒店源，multicast：组播源，subscribe：订阅源，online_search：关键字搜索；不填写则表示不指定来源，按照接口速率排序 |                   |
| recent_days            | 获取最近时间范围内更新的接口（单位天），适当减小可避免出现匹配问题                                                                                                                     | 30                |
| request_timeout        | 查询请求超时时长，单位秒(s)，用于控制查询接口文本链接的超时时长以及重试时长，调整此值能优化更新时间                                                                                                   | 10                |
| sort_timeout           | 单个接口测速超时时长，单位秒(s)；数值越大测速所属时间越长，能提高获取接口数量，但质量会有所下降；数值越小测速所需时间越短，能获取低延时的接口，质量较好；调整此值能优化更新时间                                                             | 10                |
| source_file            | 模板文件路径                                                                                                                                                | config/demo.txt   |
| subscribe_num          | 结果中偏好的订阅源接口数量                                                                                                                                         | 10                |
| urls_limit             | 单个频道接口数量                                                                                                                                              | 10                |

CHANGELOG.md
# 更新日志（Changelog）

## v1.5.6

### 2024/12/17

- ❤️ 推荐关注微信公众号（Govin），订阅更新通知与使用技巧等文章推送，还可进行答疑和交流讨论
- ⚠️ 本次更新涉及配置变更，以最新 `config/config.ini` 为准，工作流用户需复制最新配置至`user_config.ini`
- ✨ 新增白名单列表功能，支持自定义接口和订阅源关键字白名单，文件位于`config/whitelist.txt`，工作流用户为了避免冲突覆盖，建议文件重命名添加
  `user_`前缀（#584,#599）
- ✨ 新增黑名单列表功能，支持接口关键字黑名单，文件位于`config/blacklist.txt`，工作流用户为了避免冲突覆盖，建议文件重命名添加
  `user_`前缀
- ✨ 新增订阅源列表功能，文件位于`config/subscribe.txt`，工作流用户为了避免冲突覆盖，建议文件重命名添加`user_`前缀
- ✨ 新增支持获取接口速率、最低速率过滤（`open_filter_speed`、`min_speed`）
- ✨ 新增支持修改Docker服务端口环境变量（`APP_PORT`）（#619）
- ✨ 新增jsdelivr代理地址，支持TLSv1.1 和 TLSv1.2 协议（#639）
- ✨ 新增离线数据和网络数据查询开关（`open_use_cache`, `open_request`）
- ✨ 新增控制是否使用离线数据和网络数据查询（`open_use_cache`、`open_request`）
- ✨ 新增支持跳过检查是否支持ipv6（`ipv6_support`）
- ✨ 调整GUI界面布局，新增测速设置页面，跳转编辑白/黑名单、订阅源列表文本
- 🐛 修复部分m3u8接口测速导致任务超时（#621）
- 🐛 修复GUI日志线程占用问题（#655）
- 🐛 补充显示更新时间配置文档（#622）
- 🪄 优化接口测速方法，移除`yt-dlp`（#621）
- 🗑️ 移除配置：`open_ffmpeg`、`subscribe_urls`、`resolution_weight`、`response_time_weight`、`url_keywords_blacklist`

<details>
  <summary>English</summary>

- ❤️ Recommend following the WeChat public account (Govin) to subscribe to update notifications and articles on usage
  tips, as well as for Q&A and discussion.
- ⚠️ This update involves configuration changes. Refer to the latest `config/config.ini`. Workflow users need to copy
  the latest configuration to `user_config.ini`.
- ✨ Added whitelist feature, supporting custom interface and subscription source keyword whitelists. The file is located
  at `config/whitelist.txt`. To avoid conflict, workflow users are advised to rename the file with a `user_` prefix (
  #584, #599).
- ✨ Added blacklist feature, supporting interface keyword blacklists. The file is located at `config/blacklist.txt`. To
  avoid conflict, workflow users are advised to rename the file with a `user_` prefix.
- ✨ Added subscription source list feature. The file is located at `config/subscribe.txt`. To avoid conflict, workflow
  users are advised to rename the file with a `user_` prefix.
- ✨ Added support for fetching interface speed and minimum speed filtering (`open_filter_speed`, `min_speed`).
- ✨ Added support for modifying Docker server port environment variable (`APP_PORT`) (#619).
- ✨ Added jsdelivr proxy address, supporting TLSv1.1 and TLSv1.2 protocols (#639).
- ✨ Added switches for offline data and network data queries (`open_use_cache`, `open_request`).
- ✨ Added control for whether to use offline data and network data queries (`open_use_cache`, `open_request`).
- ✨ Added support for skipping the check for IPv6 support (`ipv6_support`).
- ✨ Adjusted GUI layout, added speed test settings page, and links to edit whitelist/blacklist and subscription source
  list text files.
- 🐛 Fixed issue where some m3u8 interface speed tests caused task timeouts (#621).
- 🐛 Fixed GUI log thread occupation issue (#655).
- 🐛 Added display of update time in configuration documentation (#622).
- 🪄 Optimized interface speed test method, removed `yt-dlp` (#621).
- 🗑️ Removed configurations: `open_ffmpeg`, `subscribe_urls`, `resolution_weight`, `response_time_weight`,
  `url_keywords_blacklist`.

</details>

## v1.5.5

### 2024/12/2

- ✨ 增加部分订阅源，移除失效源
- 🐛 调整github代理地址，解决访问失效（#603）
- 🐛 修复GUI测速阶段重复弹出窗口问题（#600）
- 🐛 修正宁夏/青海模板频道（#594）
- 🐛 修复IPv6结果为空问题
- 🪄 优化Docker测速CPU占用问题（#606）
- 🛠 调整部分默认配置

<details>
  <summary>English</summary>

- ✨ Added some subscription sources, removed invalid sources
- 🐛 Adjusted GitHub proxy address to fix access failure (#603)
- 🐛 Fixed repeated pop-up window issue during GUI speed test phase (#600)
- 🐛 Corrected Ningxia/Qinghai template channels (#594)
- 🐛 Fixed issue with empty IPv6 results
- 🪄 Optimized Docker speed test CPU usage (#606)
- 🛠 Adjusted some default configurations

</details>

## v1.5.4

### 2024/11/29

- ⚠️ Python 升级至 3.13，该版本已不支持 Win7，若有需要请使用 v1.5.3
- ⚠️ Github 仓库改名：iptv-api，使用旧接口地址请及时更换新地址
- ⚠️ Docker 新镜像仓库启用：guovern/iptv-api（旧版的 tv-driver 改为：guovern/iptv-api:latest，tv-requests 改为
  guovern/iptv-api:lite），iptv-api:latest 为完整版、iptv-api:lite 为精简版，请使用新的名称命令进行拉取，旧仓库将不再维护
- ❤️ 新增微信公众号关注途径（公众号搜索：Govin），推荐关注公众号，可订阅更新通知与使用技巧等文章推送，还可进行交流讨论
- ✨ 更换测速方法（yt-dlp），重构测速逻辑，提升准确性、稳定性与效率，减小接口切换延迟（#563）
- ✨ 新增支持 ARM v7（#562）
- ✨ 新增双结果 API 访问（ip/m3u, ip/txt）（#581）
- ✨ 新增启动 API 服务命令（pipenv run service）
- 🪄 优化 Docker 镜像大小（完整版：-25%，精简版：-66%）
- 🐛 修复部分播放器不支持的信息间隔符（#581）

<details>
  <summary>English</summary>

- ⚠️ Python has been upgraded to version 3.13, which no longer supports Win7. If needed, please use version v1.5.3.
- ⚠️ The GitHub repository has been renamed to iptv-api. If you are using the old API address, please update it to the
  new one promptly.
- ⚠️ New Docker image repository is now active: guovern/iptv-api (the old tv-driver is now guovern/iptv-api:latest, and
  tv-requests is now guovern/iptv-api:lite). iptv-api:latest is the full version, and iptv-api:lite is the lightweight
  version. Please use the new names to pull the images, as the old repository will no longer be maintained.
- ❤️ A new way to follow the WeChat official account (search for: Govin) has been added. It is recommended to follow the
  official account to subscribe to update notifications, usage tips, and engage in discussions.
- ✨ The speed measurement method has been changed to yt-dlp, and the speed measurement logic has been refactored to
  improve accuracy, stability, and efficiency, reducing interface switching delay (#563).
- ✨ Support for ARM v7 has been added (#562).
- ✨ Dual result API access (ip/m3u, ip/txt) has been added (#581).
- ✨ A command to start the API service (pipenv run service) has been added.
- 🪄 The size of the Docker image has been optimized (Full version: -25%, Lite version: -66%).
- 🐛 Fixed the information delimiter issue for some players that do not support it (#581).

</details>

## v1.5.3

### 2024/11/19

⚠️ 这将是支持 Win7 的最后一个版本

- 🐛 修复 GUI “显示无结果分类”设置后保存失败（#564）
- 🐛 修复命令行启动报错 (#567）

<details>
  <summary>English</summary>

⚠️ This will be the last version supporting Win7


utils/config.py
import configparser
import os
import re
import shutil
import sys


def resource_path(relative_path, persistent=False):
    """
    Get the resource path
    """
    base_path = os.path.abspath(".")
    total_path = os.path.join(base_path, relative_path)
    if persistent or os.path.exists(total_path):
        return total_path
    else:
        try:
            base_path = sys._MEIPASS
            return os.path.join(base_path, relative_path)
        except Exception:
            return total_path


def get_resolution_value(resolution_str):
    """
    Get resolution value from string
    """
    pattern = r"(\d+)[xX*](\d+)"
    match = re.search(pattern, resolution_str)
    if match:
        width, height = map(int, match.groups())
        return width * height
    else:
        return 0


class ConfigManager:

    def __init__(self):
        self.config = None
        self.load()

    def __getattr__(self, name, *args, **kwargs):
        return getattr(self.config, name, *args, **kwargs)

    @property
    def open_service(self):
        return self.config.getboolean("Settings", "open_service", fallback=True)

    @property
    def open_update(self):
        return self.config.getboolean("Settings", "open_update", fallback=True)

    @property
    def open_use_cache(self):
        return self.config.getboolean("Settings", "open_use_cache", fallback=True)

    @property
    def open_request(self):
        return self.config.getboolean("Settings", "open_request", fallback=False)

    @property
    def open_filter_speed(self):
        return self.config.getboolean(
            "Settings", "open_filter_speed", fallback=True
        )

    @property
    def open_filter_resolution(self):
        return self.config.getboolean(
            "Settings", "open_filter_resolution", fallback=True
        )

    @property
    def ipv_type(self):
        return self.config.get("Settings", "ipv_type", fallback="全部").lower()

    @property
    def open_ipv6(self):
        return (
                "ipv6" in self.ipv_type or "all" in self.ipv_type or "全部" in self.ipv_type
        )

    @property
    def ipv_type_prefer(self):
        return [
            type.strip().lower()
            for type in self.config.get(
                "Settings", "ipv_type_prefer", fallback="auto"
            ).split(",")
        ]

    @property
    def ipv4_num(self):
        return self.config.getint("Settings", "ipv4_num", fallback=15)

    @property
    def ipv6_num(self):
        return self.config.getint("Settings", "ipv6_num", fallback=15)

    @property
    def ipv6_support(self):
        return self.config.getboolean("Settings", "ipv6_support", fallback=False)

    @property
    def ipv_limit(self):
        return {
            "ipv4": self.ipv4_num,
            "ipv6": self.ipv6_num,
        }

    @property
    def origin_type_prefer(self):
        return [
            origin.strip().lower()
            for origin in self.config.get(
                "Settings",
                "origin_type_prefer",
                fallback="",
            ).split(",")
            if origin.strip().lower()
        ]

    @property
    def hotel_num(self):
        return self.config.getint("Settings", "hotel_num", fallback=10)

    @property
    def multicast_num(self):
        return self.config.getint("Settings", "multicast_num", fallback=10)

    @property
    def subscribe_num(self):
        return self.config.getint("Settings", "subscribe_num", fallback=10)

    @property
    def online_search_num(self):
        return self.config.getint("Settings", "online_search_num", fallback=10)

    @property
    def source_limits(self):
        return {
            "hotel": self.hotel_num,
            "multicast": self.multicast_num,
            "subscribe": self.subscribe_num,
            "online_search": self.online_search_num,
        }

    @property
    def min_speed(self):
        return self.config.getfloat("Settings", "min_speed", fallback=0.5)

    @property
    def min_resolution(self):
        return self.config.get("Settings", "min_resolution", fallback="1920x1080")

    @property
    def min_resolution_value(self):
        return get_resolution_value(self.min_resolution)

    @property
    def urls_limit(self):
        return self.config.getint("Settings", "urls_limit", fallback=30)

    @property
    def open_url_info(self):
        return self.config.getboolean("Settings", "open_url_info", fallback=True)

    @property
    def recent_days(self):
        return self.config.getint("Settings", "recent_days", fallback=30)

    @property
    def source_file(self):
        return self.config.get("Settings", "source_file", fallback="config/demo.txt")

    @property
    def final_file(self):
        return self.config.get("Settings", "final_file", fallback="output/result.txt")

    @property
    def open_m3u_result(self):
        return self.config.getboolean("Settings", "open_m3u_result", fallback=True)

    @property
    def open_keep_all(self):
        return self.config.getboolean("Settings", "open_keep_all", fallback=False)

    @property
    def open_subscribe(self):
        return self.config.getboolean("Settings", f"open_subscribe", fallback=True)

    @property
    def open_hotel(self):
        return self.config.getboolean("Settings", f"open_hotel", fallback=True)

    @property
    def open_hotel_fofa(self):
        return self.config.getboolean("Settings", f"open_hotel_fofa", fallback=True)

    @property
    def open_hotel_foodie(self):
        return self.config.getboolean("Settings", f"open_hotel_foodie", fallback=True)

    @property
    def open_multicast(self):
        return self.config.getboolean("Settings", f"open_multicast", fallback=True)

    @property
    def open_multicast_fofa(self):
        return self.config.getboolean("Settings", f"open_multicast_fofa", fallback=True)

    @property
    def open_multicast_foodie(self):
        return self.config.getboolean(
            "Settings", f"open_multicast_foodie", fallback=True
        )

    @property
    def open_online_search(self):
        return self.config.getboolean("Settings", f"open_online_search", fallback=True)

    @property
    def open_method(self):
        return {
            "subscribe": self.open_subscribe,
            "hotel": self.open_hotel,
            "multicast": self.open_multicast,
            "online_search": self.open_online_search,
            "hotel_fofa": self.open_hotel_fofa,
            "hotel_foodie": self.open_hotel_foodie,
            "multicast_fofa": self.open_multicast_fofa,
            "multicast_foodie": self.open_multicast_foodie,
        }

    @property
    def open_use_old_result(self):
        return self.config.getboolean("Settings", "open_use_old_result", fallback=True)

    @property
    def open_sort(self):
        return self.config.getboolean("Settings", "open_sort", fallback=True)

    @property
    def open_update_time(self):
        return self.config.getboolean("Settings", "open_update_time", fallback=True)

    @property
    def multicast_region_list(self):
        return [
            region.strip()
            for region in self.config.get(
                "Settings", "multicast_region_list", fallback="全部"
            ).split(",")
            if region.strip()
        ]

    @property
    def hotel_region_list(self):
        return [
            region.strip()
            for region in self.config.get(
                "Settings", "hotel_region_list", fallback="全部"
            ).split(",")
            if region.strip()
        ]

    @property
    def request_timeout(self):
        return self.config.getint("Settings", "request_timeout", fallback=10)

    @property
    def sort_timeout(self):
        return self.config.getint("Settings", "sort_timeout", fallback=10)

    @property
    def open_proxy(self):
        return self.config.getboolean("Settings", "open_proxy", fallback=False)

    @property
    def open_driver(self):
        return not os.environ.get("LITE") and self.config.getboolean(
            "Settings", "open_driver", fallback=True
        )

    @property
    def hotel_page_num(self):
        return self.config.getint("Settings", "hotel_page_num", fallback=1)

    @property
    def multicast_page_num(self):
        return self.config.getint("Settings", "multicast_page_num", fallback=1)

    @property
    def online_search_page_num(self):
        return config.getint("Settings", "online_search_page_num", fallback=1)

    @property
    def open_empty_category(self):
        return self.config.getboolean("Settings", "open_empty_category", fallback=True)

    @property
    def app_port(self):
        return os.environ.get("APP_PORT") or self.config.getint("Settings", "app_port", fallback=8000)

    def load(self):
        """
        Load the config
        """
        self.config = configparser.ConfigParser()
        user_config_path = resource_path("config/user_config.ini")
        default_config_path = resource_path("config/config.ini")

        config_files = [user_config_path, default_config_path]
        for config_file in config_files:
            if os.path.exists(config_file):
                with open(config_file, "r", encoding="utf-8") as f:
                    self.config.read_file(f)
                break

    def set(self, section, key, value):
        """
        Set the config
        """
        self.config.set(section, key, value)

    def save(self):
        """
        Save config with write
        """
        user_config_file = "config/" + (
            "user_config.ini"
            if os.path.exists(resource_path("user_config.ini"))
            else "config.ini"
        )
        user_config_path = resource_path(user_config_file, persistent=True)
        if not os.path.exists(user_config_path):
            os.makedirs(os.path.dirname(user_config_path), exist_ok=True)
        with open(user_config_path, "w", encoding="utf-8") as configfile:
            self.config.write(configfile)

    def copy(self):
        """
        Copy config files to current directory
        """
        dest_folder = os.path.join(os.getcwd(), "config")
        try:
            src_dir = resource_path("config")
            if os.path.exists(src_dir):
                if not os.path.exists(dest_folder):
                    os.makedirs(dest_folder, exist_ok=True)

                for root, _, files in os.walk(src_dir):
                    for file in files:
                        src_file_path = os.path.join(root, file)
                        relative_path = os.path.relpath(src_file_path, src_dir)
                        dest_file_path = os.path.join(dest_folder, relative_path)

                        dest_file_dir = os.path.dirname(dest_file_path)
                        if not os.path.exists(dest_file_dir):
                            os.makedirs(dest_file_dir, exist_ok=True)

                        if not os.path.exists(dest_file_path):
                            shutil.copy(src_file_path, dest_file_path)
        except Exception as e:
            print(f"Failed to copy files: {str(e)}")


config = ConfigManager()
